### ISSUE TITLE
UnicodeDecodeError for non-utf8 encoded bytes

### ISSUE THREAD

**Issue by TheElementalOfDestruction** [seamustuohy/RTFDE#7]
```
  File "C:\Program Files\Python39\lib\site-packages\extract_msg\message_base.py", line 206, in deencapsulatedRtf
    self._deencapsultor = RTFDE.DeEncapsulator(self.rtfBody)
  File "C:\Program Files\Python39\lib\site-packages\RTFDE\deencapsulate.py", line 69, in __init__
    self.raw_rtf = raw_rtf.decode()
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 22090: invalid start byte
```

Looks like the deencapsulator handles bytes by just trying to plainly decode them with no regards to the content. Unfortunately, some of the bodies cannot be utf-8 decoded as I have just found out. Specifically, if your body is using ANSI codepage 1252 and contains an ellipses, this will be written to the body with the code `'\x85'`, which utf-8 cannot decode. As such, it looks like smarter decoding may be needed for bytes.

**Comment by TheElementalOfDestruction** [comment]
I do note that it does feel like a violation of the standard (which it actually looks like outlook itself violated it) where an 8bit character is not encoded as `\'XX` but I'm looking into it for more details.

**Comment by TheElementalOfDestruction** [comment]
Additionally, it looks like the `\binN` field can have *any* 8 bit character without being a violation, and should be treated as binary data, not text. Looks like that definitely needs some kind of transformation to strip those out when deencapsulating. Here is the documentation relevant to `\binN` for convenience:
![image](https://user-images.githubusercontent.com/24580325/173210797-c00f9be3-46f1-46ab-997c-f9b44e692159.png)
![image](https://user-images.githubusercontent.com/24580325/173210857-dd6cf16b-7c07-4c8b-9d45-d2f52ff36f0d.png)
(Of course, while this is listed under picture data, it's not reserved for it, as other parts of the doc reference it)

Looks like if you see a \bin keyword, it should have it's value read, look at the next character to see if it is a space, drop that character if it is a space, then drop the next N characters (including that one if it was not a space).

Also the first image says anything from word (which is typically what encapsulated data is marked as being generated from) should expect 8 bit characters to be present. While actually using them seems like it would not particularly be backwards compatible, it's not a complete violation, if my understanding is right.

**Comment by seamustuohy** [comment]
@TheElementalOfDestruction I believe I've addressed this issue in the [major_structural_changes](https://github.com/seamustuohy/RTFDE/tree/major_structural_changes) branch.  If you have a test file you should be able to verify by using the ["private data test folder" functionality described here](https://github.com/seamustuohy/RTFDE/tree/major_structural_changes/tests/test_data/personal_rtf_test_files).

**Comment by TheElementalOfDestruction** [comment]
I'll run some tests when I have the time to make sure these are all working correctly on my end. I'll check for exceptions from any of my files first before checking that the output all looks good. I'll let you know if I have any more issues.

**Comment by TheElementalOfDestruction** [comment]
Currently seem to be getting the same error on every file I try including files that previously worked. Looks like the second item in the loop is a Tree instance but it's data field is the string "document" which breaks the `is_hexarray` function. 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 1122, in htmlBody
    self._htmlBody = self.deencapsulateBody(self.rtfBody, DeencapType.HTML)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 192, in deencapsulateBody
    if self.deencapsulatedRtf and self.deencapsulatedRtf.content_type == 'html':
       ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 990, in deencapsulatedRtf
    self._deencapsultor.deencapsulate()
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\deencapsulate.py", line 113, in deencapsulate
    Decoder.update_children(self.full_tree)
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\text_extraction.py", line 614, in update_children
    obj.children = [i for i in self.iterate_on_children(children)]
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\text_extraction.py", line 614, in <listcomp>
    obj.children = [i for i in self.iterate_on_children(children)]
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\text_extraction.py", line 663, in iterate_on_children
    elif is_hexarray(item):
         ^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\text_extraction.py", line 543, in is_hexarray
    if item.data.value == 'hexarray':
       ^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'value'
```

**Comment by TheElementalOfDestruction** [comment]
Decided to look at lark since that's what is creating the object, and it looks like `data` on a Tree is always meant to be a string since it's a name. Might be that `value` wasn't supposed to be there?

**Comment by seamustuohy** [comment]
Oh! Thank you. Ill try to recreate this tonight. I imagine this might be a
lark-parser version incompatibility since I did various tests with the
bleeding edge version.

**Comment by seamustuohy** [comment]
Can you run `pip3 list | grep lark` and provide the output? I want to make sure I'm testing this on the correct releases of lark. I currently can't recreate it on 1.1.5.

**Comment by TheElementalOfDestruction** [comment]
I don't have grep on this machine, though I knew what I was expecting from the output and got 1.1.5, so I double checked... no error. I'm honestly baffled so I went and double checked my version number from before and despite having just installed the new version of the module, it was *not* 1.1.5 in the code but *was* 1.1.5 according to pip. So looks like something completely stupid happened there with pip. Sorry about that 😅 

I am getting new errors after verifying everything in my package list is correct and reinstalling the new version of RTFDE.

```
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 1121, in htmlBody
    self._htmlBody = self.deencapsulateBody(self.rtfBody, DeencapType.HTML)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 192, in deencapsulateBody
    if self.deencapsulatedRtf and self.deencapsulatedRtf.content_type == 'html':
       ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 989, in deencapsulatedRtf
    self._deencapsultor.deencapsulate()
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\deencapsulate.py", line 117, in deencapsulate
    htmlrtf_stripped = self.strip_htmlrtf_tokens()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\deencapsulate.py", line 244, in strip_htmlrtf_tokens
    deleter = DeleteTokensFromTree(tokens_to_delete)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\transformers.py", line 297, in __init__
    self.delete_start_pos = {i.start_pos for i in self.to_delete}
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\transformers.py", line 297, in <setcomp>
    self.delete_start_pos = {i.start_pos for i in self.to_delete}
                             ^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'start_pos'
```
and
```
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 1121, in htmlBody
    self._htmlBody = self.deencapsulateBody(self.rtfBody, DeencapType.HTML)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 192, in deencapsulateBody
    if self.deencapsulatedRtf and self.deencapsulatedRtf.content_type == 'html':
       ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\extract_msg\message_base.py", line 966, in deencapsulatedRtf
    self._deencapsultor = RTFDE.DeEncapsulator(body)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Destruction\AppData\Roaming\Python\Python311\site-packages\RTFDE\deencapsulate.py", line 91, in __init__
    self.raw_rtf: str = raw_rtf.decode()
                        ^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 22080: invalid start byte
```
Which is just effectively the original error that this issue is, which makes it seem like it is still just trying to decode the bytes as UTF-8.

Both only happen on select files, but the first is the one I don't really understand, so I'll try and see if I can figure out what triggers it if you need me to.

**Comment by seamustuohy** [comment]
The former one is erroring out because one of the tokens it is looking to delete has already been transformed into a string. If you could replace line 297 in `RTFDE\transformers.py` with the below code and then tell me what that string characters are printed is it would help me track down where the accidental early transformation is and to fix it. (I'm guessing it's whitespace. It needs such careful handling all over the place.)

```
           try:
               self.delete_start_pos = {i.start_pos for i in self.to_delete}
           except AttributeError as _e:
               print([i for i in tokens_to_delete if isinstance(i,str)])
               raise _e
```


On the latter I'll want to know what the encoding of those files is (not the RTF in the file, but the file itself.) If your system has `file` on it `file -i FILENAME.rtf` should tell you the charset being used in the file. That should get me started.

**Comment by TheElementalOfDestruction** [comment]
I'll run that test code later to see what it outputs. As for the other file, the result was `test.rtf: Rich Text Format data, version 1, ANSI, code page 1252, default middle east language ID 1025`. TBH it *should* be written as `\'85` but my understanding of RTF standard is that it technically isn't mandatory. Also this one randomly has a null character at the end of it.

**Comment by TheElementalOfDestruction** [comment]
Tried to run your code real quick, but I believe Token is a subclass of string since it output a lot of those. I'll have to modify the test code to look for not a subclass

**Comment by seamustuohy** [comment]
Sorry, that's my bad use this `isinstance` line instead to rule that out. 

```
print([i for i in tokens_to_delete if not isinstance(i,Token) is False])
```

**Comment by seamustuohy** [comment]
`0x85` is the NEL newline that I didn't think was used anymore. A seach says it was used in mainframes so I imagine some legacy systems might still be using it. But, the issues with non linux newline characters being used has been a thing before which is why it replaces all CR and CRLF newlines with a single LF `\n` char. I'm headed out right now. But, I'll try to add extra conversations for all possible newline chars and see if that address' it.

**Comment by TheElementalOfDestruction** [comment]
Where are you seeing that 0x85 is the NEL character in windows 1252? Wikipedia and word both tell me it's ellipsis

**Comment by seamustuohy** [comment]
Well, the place where we are seeing the error is before any decoding, or
parsing of the RTF text happens. It's just decoding the outlying
encapsulated RTF blob. I could be wrong, but I think file identifies the
code page and default encoding (ANSO) of the data within the RTF file; not
the overarching file. But, I could be wrong. NEL is just 0x85 in Unicode.
If the overarching RTF twct is also encoded in a format other than ASCII or
UTF-8 that would be new to me and id ha e to check it out.

**Comment by TheElementalOfDestruction** [comment]
The character is 0x85 in the binary, embedded in the text as a plain character. It's referencing that character in ansi windows 1252, since RTF files don't actually contain Unicode character plainly.

**Comment by TheElementalOfDestruction** [comment]
The Word 2007 RTF standard has the following lines that should help clarify the proper way to deal with the RTF data:
 > RTF files are usually 7-bit ASCII plain text, consisting of control words, control symbols, and groups. RTF files are easily transmitted between most PC based operating systems because of their 7-bit ASCII characters. However, converters that communicate with Microsoft Word for Windows or Microsoft Word for the Macintosh should expect data transfer as 8-bit characters and binary data (see \binN) can contain any 8-bit values.

> If the character is anything other than an opening brace ({), closing brace (}), backslash (\), or a CRLF
(carriage return/line feed), the reader assumes that the character is plain text and writes the character
to the current destination using the current formatting properties

> In general RTF should be written out with all characters above 0x7F in the escaped form, `\'hh` or the `\uN` form if an ANSI version doesn’t exist.

This means a few things. The first is that regardless of the encoding, any *encoding independent* 8-bit character can appear, including ones that don't exist in the current encoding. If you preempt decoding by stripping out those binary sections though, that shouldn't cause issues.

The second thing this means is that the specification explicitly uses "should" instead of "must" for saying that non 7-bit characters should escaped. If an ANSI version of the character exists, then it would not be a violation of the current standard to write it directly instead of escaping it. In the case of the file that is causing an error, this is exactly what has happened: the character was written without escaping it, since it is included in the codepage.

Is it *good* that that is acceptable in the standard? I'd say no, but it's what we have to deal with. Stuff like this is actually why my little RTF handler built-in to extract_msg just reads everything as bytes, lol.

**Comment by TheElementalOfDestruction** [comment]
Additionally, I finally got what string caused the problem, and it's literally just an empty string. `['']`

**Comment by seamustuohy** [comment]
Well, it sounds like I need to follow your approach and just deal with everything as bytes then.

**Comment by seamustuohy** [comment]
@TheElementalOfDestruction I'm converting this to work with everything as bytes. I'd like your thoughts on my approach before I start polishing it. 
 
* I'm extracting out all `\\bin` data. I'm making it available to the user as an Attribute in case they want to make use of it. But, converting that binary back is out of scope.
* I'm treating non-7-bit characters which are included in a file as not part of the encapsulation. As such, I plan to leave them alone and pass them back with the bytes I return. All the other characters are properly decoded from their code-page and encoded into python3 string bytes. But, my current plan is to leave any raw bytes untouched. This would mean that those raw bytes would be included in any deencapsulated RTF data that is returned. This would leave it up to the user to extract them if they wanted to fully decode the html.

**Comment by TheElementalOfDestruction** [comment]
If they need the binary data, it would need to be stored in a list of sections, since the bin sections are all going to be independent of each other. If the user can use something about the token or whatever to actually find the correct bin statement, and everything about that would sound good.

For the second, I'd have to check the documentation on how encapsulated HTML should be dealt with, but it *may* be possible that the HTML identifies its encoding differently than the file, in which case you may end up with a character that is encoded wrong. I don't think this is a bad approach though, since the user should have no real issue parsing the HTML through something like BS4 to check what encoding the bytes are supposed to be, checking the RTF encoding, and then doing a bit of magic to correct any potential issues that arise from a mismatch. And if mismatches like that are common, code like that could potentially just be integrated back up to RTFDE through a pull request or something.

Overall I think that's a pretty good approach. I'll double check the documentation and my example files when I have time though.

**Comment by seamustuohy** [comment]
OK, I've made the change over in the [lets_go_bytes branch](https://github.com/seamustuohy/RTFDE/tree/lets_go_bytes). This includes all the other major changes. Take a look and see if that address' the issues you have come across.

**Comment by TheElementalOfDestruction** [comment]
Alright, some news, it seems to be caused by the unicode surrogates or something. Here is what I see in the file, and here is what I see when printing out information about the tokens to be deleted:
```rtf
\u-10179 ?\u-8694 ?\htmlrtf0
```
```
b'\xf0\x9f\x98\x8a' 3347
<class 'bytes'> b''
b'\\htmlrtf0 ' 3366
```
The first line in the output is that result of combining the two surrogates, the next line is the empty token that was found, and the line after that is the token immediately after the question mark. I can tell what is what by the different way I had each item print (forcibly ignoring the errors that occurred to print out everything). Things printed with a number after them are instances of Token while things printed as a class before an entry are not instances of Token.

Knowing this, I managed to create a pretty minimal example file to use for testing against. The following set of RTF data will cause the error:

```rtf
{\rtf1\ansi\ansicpg1252\fromhtml1 \fbidis \deff0{\fonttbl
{\f0\fswiss\fcharset0 Arial;}{\f1\fmodern Courier New;}
{\f2\fnil\fcharset2 Symbol;}
{\f3\fmodern\fcharset0 Courier New;}
{\f4\fswiss\fcharset0 "Segoe UI Emoji";}
{\f5\fswiss\fcharset0 "Century Gothic";}}
{\colortbl\red0\green0\blue0;\red5\green99\blue193;}
\uc1\pard\plain
{\*\htmltag84 &#128522;}\htmlrtf \u-10179 ?\u-8694 ?\htmlrtf0
}
```
Putting this all together into a little bit of minimal Python code, this code snippet will also cause the error to be thrown.
```py
import RTFDE

a = RTFDE.DeEncapsulator(b'{\\rtf1\\ansi\\ansicpg1252\\fromhtml1 \\fbidis \\deff0{\\fonttbl{\\f0\\fswiss\\fcharset0 Arial;}{\\f1\\fmodern Courier New;}{\\f2\\fnil\\fcharset2 Symbol;}{\\f3\\fmodern\\fcharset0 Courier New;}{\\f4\\fswiss\\fcharset0 "Segoe UI Emoji";}{\\f5\\fswiss\\fcharset0 "Century Gothic";}}{\\colortbl\\red0\\green0\\blue0;\\red5\\green99\\blue193;}\\uc1\\pard\\plain{\\*\\htmltag84 &#128522;}\\htmlrtf \\u-10179 ?\\u-8694 ?\\htmlrtf0}')
a.deencapsulate() # Error thrown here.
```

Hope all of this will help you track down the error, though I may look around a bit myself to see if I can identify what creates it, given I now have a rather small file to test against.

**Comment by TheElementalOfDestruction** [comment]
I believe I may have tracked down the source of the issue. In `text_extraction.py` there is a line that will simple replace one of the tokens with an empty binary string when combining surrogates. I believe that it was *supposed* to do was replace it with an empty Token or something to indicate it's deletion without changing it from being a Token. This can be seen in the following line:

https://github.com/seamustuohy/RTFDE/blob/e34ad3ae043fb74cfc92d843109767a58bf57136/RTFDE/text_extraction.py#L495

**Comment by seamustuohy** [comment]
Thanks! I've pushed a quick update to ensure that it pushes a token instead of a bytes string. It address' that bug when I test it on the example file you provided.

**Comment by TheElementalOfDestruction** [comment]
Good thing I let my tests run, cause I got another file that breaks things in a weird way. Two surrogates written like this with some text after it are parsing in a way that doesn't seem entirely correct:
```rtf
\u55357?\u56542? +sometest
```
The second question mark is, for whatever reason, being put into a tree with the next string section, and seems to be causing some sort of issue later since when parsing it actually straight up disappears while the first one is still around. Eventually, I see that the surrogates aren't combined leading to a unicode decode error. I'm trying to see if I can figure out why the code is doing this, but I need to create a small example to test against.

**Comment by seamustuohy** [comment]
Thank you! If you create the test case I'll happily track down the issue.
The surrogates code is a special little circle of hell, because I've yet to
implement special logging like with other sections. But, I know it well so
it should be pretty easy to track down with a test file.

**Comment by TheElementalOfDestruction** [comment]
Turns out it appears to be an issue with that specific high surrogate, since simply changing the value on the previous test file threw the error I was getting. Here is the new RTF that will produce the error, though I want to see if I might be able to figure out the cause too.

```rtf
{\rtf1\ansi\ansicpg1252\fromhtml1 \fbidis \deff0{\fonttbl
{\f0\fswiss\fcharset0 Arial;}{\f1\fmodern Courier New;}
{\f2\fnil\fcharset2 Symbol;}
{\f3\fmodern\fcharset0 Courier New;}
{\f4\fswiss\fcharset0 "Segoe UI Emoji";}
{\f5\fswiss\fcharset0 "Century Gothic";}}
{\colortbl\red0\green0\blue0;\red5\green99\blue193;}
\uc1\pard\plain
{\*\htmltag84 &#128522;}\htmlrtf \u55357 ?\u56542 ?\htmlrtf0
}
```

**Comment by TheElementalOfDestruction** [comment]
Ah, I see the problem. For whatever reason, outlook has decided here that it wants to write unicode characters using *not* a 16 bit signed integer. *Technically* this seems valid within the standard, as readers must allow for up to 10 digits optionally preceded by a minus sign (according to the docs) but it's rather stupid for them to just ignore typical writing like that.

**Comment by seamustuohy** [comment]
OK, I think I've addressed parsing the non-16 bit unicode chars in the [lets_go_bytes](https://github.com/seamustuohy/RTFDE/tree/lets_go_bytes] branch. I had to make a separate test file because the one provided removed the surrogates because they were placed in an `\\htmlrtf` block. All text within `\\htmlrtf` blocks is removed from the content. Take a look and let me know.

**Comment by TheElementalOfDestruction** [comment]
I ran it through my battery of tests (6.7k files) and I am down to *one* file having an issue, and it seems to be caused by a Unicode character being followed by a `\'HH` control, which causes a hex array tree to form which ends up becoming empty, which later causes a nonetype has no attribute decode error. I managed to make a brief example that shows this, and from what I can tell it looks to be valid RTF.

```
{\rtf1\ansi\ansicpg1252\fromhtml1 \fbidis \deff0{\fonttbl
{\f0\fswiss\fcharset0 Arial;}{\f1\fmodern Courier New;}
{\f2\fnil\fcharset2 Symbol;}
{\f3\fmodern\fcharset0 Courier New;}
{\f4\fswiss\fcharset0 "Segoe UI Emoji";}
{\f5\fswiss\fcharset0 "Century Gothic";}}
{\colortbl\red0\green0\blue0;\red5\green99\blue193;}
\uc1\pard\plain
{\*\htmltag84 &#128522;}\htmlrtf \u55357\'20\u56542 ?\htmlrtf0
}
```
This is literally the last file that is having an error, so none of my files show anything wrong once this is fixed.

Sorry for the wait, you responded right as I got stick (I still am, but I wanted to get some work done to get my mind back up and running, so here we are).

**Comment by seamustuohy** [comment]
Addressed in the 0.1.0 version which is in the final notification and testing stage on [dev](https://github.com/seamustuohy/RTFDE/commit/1b29704a47e6c3a118ebb3ed2ac96e640312f8fb)

**Comment by TheElementalOfDestruction** [comment]
Yep, my final test file is now passing just fine, so looks like everything is good. I did notice though that it looks like a debug print got left enabled at test_extraction.py line 709, but since it's on the dev branch that might have been intentional. 

I also reran my entire set of tests and all of them are passing from what I can see, so great job.
